### PR TITLE
ACDD 1.3 rebase against current master

### PR DESCRIPTION
**Pull request by benjwadams**
ACDD-1.3 support.  Hopefully the last PR needed in light of some earlier issues which came up.

Supercedes https://github.com/ioos/compliance-checker/pull/253 due to overcoming merge conflicts on master.

Also removed "2.2.0" `__version__` attribute until we are ready to formally declare a 2.2.0 release and bump the version up.

Implements #127, #133, #186

### CODE REVIEW

File: compliance_checker/acdd.py [hunk 1]
```diff
@@ -384,6 +451,237 @@ def check_time_extents(self, ds):
...
+                'program',
+                'publisher_institution',
+            ])
+
+    def check_history(self, ds):
+        #@TODO Create a history check
+        return
+
+    def check_source(self, ds):
+        #@TODO Create a source check
```
Comment (lukecampbell) [inline compliance_checker/acdd.py]: Could you do a simple check here to ensure that the attribute exists and is not empty?

**Comment by lukecampbell** [comment]
Can you please check for the existence of "acknowledgement" and/or "acknowledgment".   ?

ACDD references "acknowledgement" (with an e) but commonly datasets use "acknowledgment"

> In dictionaries, the preferred spelling is listed first among the headwords of an entry. Examples follow: acknowledgement vs acknowledgment: acknowledgement is preferred in British English, acknowledgment in American English.

File: compliance_checker/acdd.py (hunk 1 above)
Comment (benjwadams) [inline compliance_checker/acdd.py]: There already exists checks to see if these are present.  Don't think it checks if empty or not.

**Comment by lukecampbell** [comment]
Also the cdm_data_type is no longer required to be one of those ISO defined types, they are now a THREDDS dataType

> cdm_data_type	The data type, as derived from Unidata's Common Data Model Scientific Data types and understood by THREDDS. (This is a THREDDS "dataType", and is different from the CF NetCDF attribute 'featureType', which indicates a Discrete Sampling Geometry file in CF.)

**Comment by benjwadams** [comment]
Source of the above?

**Comment by lukecampbell** [comment]
http://wiki.esipfed.org/index.php/Attribute_Convention_for_Data_Discovery_1-3

https://github.com/Unidata/thredds/issues/72

```
public enum ucar.nc2.constants.FeatureType {
  ANY, // no specific type

  GRID,
  RADIAL,
  SWATH,
  IMAGE,

  ANY_POINT, // any of the following
  POINT,
  PROFILE,
  SECTION,
  STATION,
  STATION_PROFILE,
  TRAJECTORY;
}
```

**Comment by lukecampbell** [comment]
Just leave it as is, it should be deprecated anyway.

https://github.com/Unidata/thredds/issues/72#issuecomment-57932413

File: compliance_checker/acdd.py (hunk 1 above)
Comment (benjwadams) [inline compliance_checker/acdd.py]: Going to modify *all* presence checks to see if the attribute value is empty and fail if it is.

**Comment by benjwadams** [comment]
Per the acknowledgment/acknowledgement issue, I have coded it to check both, but perhaps per the spec, we ought to fail if the non-preferred spelling is used, and make an error message to that effect?

**Comment by lukecampbell** [comment]
Let's scrap anything that's "human readable" that's purely objective and I doubt we'll ever be able to really check that in a checker.

For example:

２００６年全人口統計 is perfectly human readable but good luck on writing a checker that could read it.

**Comment by benjwadams** [comment]
@lukecampbell Above is done.

**Comment by daf** [comment]
Installing with a fresh conda environment:

```
$ conda create -n cc-acdd13 python=3 --file=requirements.txt -c conda-forge
$ source activate cc-acdd13
$ pip install -e .
$ python cchecker.py --help
```

gives

```
usage: cchecker.py [-h] --test
                   {acdd,acdd:1.1,acdd:1.3,acdd:latest,cf,cf:1.6,cf:latest,ioos,ioos:0.1,ioos:latest}
                   [{acdd,acdd:1.1,acdd:1.3,acdd:latest,cf,cf:1.6,cf:latest,ioos,ioos:0.1,ioos:latest} ...]
                   [--criteria [{lenient,normal,strict}]] [--verbose]
                   [-f {text,html,json}] [-o OUTPUT] [-V]
                   [dataset_location [dataset_location ...]]

positional arguments:
  dataset_location      Defines the location of the dataset to be checked.

optional arguments:
  -h, --help            show this help message and exit
  --test {acdd,acdd:1.1,acdd:1.3,acdd:latest,cf,cf:1.6,cf:latest,ioos,ioos:0.1,ioos:latest} [{acdd,acdd:1.1,acdd:1.3,acdd:latest,cf,cf:1.6,cf:latest,ioos,ioos:0.1,ioos:latest} ...], -t {acdd,acdd:1.1,acdd:1.3,acdd:latest,cf,cf:1.6,cf:latest,ioos,ioos:0.1,ioos:latest} [{acdd,acdd:1.1,acdd:1.3,acdd:latest,cf,cf:1.6,cf:latest,ioos,ioos:0.1,ioos:latest} ...], --test= {acdd,acdd:1.1,acdd:1.3,acdd:latest,cf,cf:1.6,cf:latest,ioos,ioos:0.1,ioos:latest} [{acdd,acdd:1.1,acdd:1.3,acdd:latest,cf,cf:1.6,cf:latest,ioos,ioos:0.1,ioos:latest} ...], -t= {acdd,acdd:1.1,acdd:1.3,acdd:latest,cf,cf:1.6,cf:latest,ioos,ioos:0.1,ioos:latest} [{acdd,acdd:1.1,acdd:1.3,acdd:latest,cf,cf:1.6,cf:latest,ioos,ioos:0.1,ioos:latest} ...]
                        Select the Checks you want to perform.
  --criteria [{lenient,normal,strict}], -c [{lenient,normal,strict}]
                        Define the criteria for the checks. Either Strict,
                        Normal, or Lenient. Defaults to Normal.
  --verbose, -v         Increase output. May be specified up to three times.
  -f {text,html,json}, --format {text,html,json}
                        Output format
  -o OUTPUT, --output OUTPUT
                        Output filename
  -V, --version         Display the IOOS Compliance Checker version
                        information.
```

Anything we can do to trim that silly list?  Is it really giving all permutations?

**Comment by daf** [comment]
Ok, some notes:

1. `--test` handling def a bit different than my previous testing.  Can't specify more than one `--test=`, or the last one takes preference.  `--test` without `=` almost always requires a `--`.  There's not much we can do about this, though, more of just a remark.

2. It would be nice if the banner text at the start of the results indicated what the test version was expanded to - right now it just echoes whatever the user gave.  This is a "nice to have"

3. I got a checker error with ACDD 1.1/1.3 on a dataset I had in my Downloads dir, which is a programming error on our part, but I just tried to find the file online again and it differs slightly and does not trigger the error.  I'll debug for a few minutes, as it's def a real issue, but need to up the file somewhere for anyone else to reproduce.

```
WARNING: The following exceptions occured during the acdd:1.1 checker (possibly indicate compliance checker issues):
Traceback (most recent call last):
  File "cchecker.py", line 67, in <module>
    sys.exit(main())
  File "cchecker.py", line 56, in main
    args.format)
  File "/home/daf/dev/compliance-checker/compliance_checker/runner.py", line 55, in run_checker
    errors_occurred = cls.check_errors(score_groups, verbose)
  File "/home/daf/dev/compliance-checker/compliance_checker/runner.py", line 141, in check_errors
    print("%s.%s: %s" % (checker, check_name, epair[0].message), file=sys.stderr)
AttributeError: 'TypeError' object has no attribute 'message'
```

Other than these I'd say this is good, thanks guys.  Let me know if you want to change anything for these or add them later.

**Comment by DanielJMaher** [comment]
@daf So 1 was a non actionable comment and 2) Is a nice to have, but not needed and 3) you resolved?

Can i give this the thumbs up merge and create an issue around 2 for the next release?

**Comment by daf** [comment]
@DanielJMaher @benjwadams merge benjwadams/compliance-checker#2 plz then this can be merged.

**Comment by benjwadams** [comment]
Per 1)  that was definitely one advantage of comma delimiting the tests -- there weren't issues with datasets being parsed as separate tests, for example.

**Comment by benjwadams** [comment]
@daf All set.  Also restored a behavior to default to 'acdd' checker if test was unspecified as current master branch behaves.

**Comment by benjwadams** [comment]
:tada: :fireworks:

**Pull request merged**